src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs

```diff
@@ -57,159 +65,237 @@ public void Free()
+        {
+            return GetEnumerator();
+        }
 
-                if (values.Length == 1)
-                {
-                    _dictionary[k] = values[0];
-                }
-                else
-                {
-                    arrayBuilder = new ArrayBuilder<V>(values.Length);
-                    arrayBuilder.AddRange(values);
-                    _dictionary[k] = arrayBuilder;
-                }
-            }
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
         }
 
         /// <summary>
-        /// Get the number of values associated with a key.
+        /// Get all values associated with K, in the order they were added.
+        /// Returns empty read-only array if no values were present.
         /// </summary>
-        public int GetCountForKey(K k)
+        public ImmutableArray<V> this[K k]
         {
-            object item;
-            if (!this.IsEmpty && _dictionary.TryGetValue(k, out item))
+            get
             {
-                return (item as ArrayBuilder<V>)?.Count ?? 1;
-            }
+                ValueSet valueSet;
+                if (!this.IsEmpty && _dictionary.TryGetValue(k, out valueSet))
+                {
+                    Debug.Assert(valueSet.Count >= 1);
+                    return valueSet.Items;
+                }
 
-            return 0;
+                return ImmutableArray<V>.Empty;
+            }
         }
 
         /// <summary>
-        /// Returns true if one or more items with given key have been added.
+        /// Get a collection of all the keys.
         /// </summary>
-        public bool ContainsKey(K k)
+        public Dictionary<K, ValueSet>.KeyCollection Keys
         {
-            return !this.IsEmpty && _dictionary.ContainsKey(k);
+            get { return this.IsEmpty ? s_emptyDictionary.Keys : _dictionary.Keys; }
         }
 
-        /// <summary>
-        /// Get all values associated with K, in the order they were added.
-        /// Returns empty read-only array if no values were present.
-        /// </summary>
-        public ImmutableArray<V> this[K k]
+        public struct ValueSet : IEnumerable<V>
         {
-            get
+            /// <summary>
+            /// Each value is either a single V or an <see cref="ArrayBuilder{V}"/>.
+            /// Never null.
+            /// </summary>
+            private readonly object _value;
+
+            // By default we allocate array builders with a size of two.  That's two store
+            // the single item already in _value, and to store the item we're adding.  
+            // In general, we presume that the amount of values per key will be low, so this
+            // means we have very little overhead when there are multiple keys per value.
+            private static ObjectPool<ArrayBuilder<V>> s_builderPool = new ObjectPool<ArrayBuilder<V>>(
```

Comment:
Isn't there a pooled version of array builder? e.g. `ArrayBuilder<V>.GetInstance()`